### PR TITLE
Add an opportunity to set Signed Headers from GET parameters in signature v4

### PR DESCRIPTION
**Pull request by mikhalenia**
Fix the problem with signature v4. It has to be an opportunity to send some Signed Header values through GET parameter.

## Description

I work on the project which uses aws-sdk v2.352.0 on Node.js. We have some case when we need to send pre-signed request to S3. When we use the real amazon S3 it works fine. But when I try to switch it to MinIO we have two problems:
1) There were headers present in the request which were not signed.
2) The request signature we calculated does not match the signature you provided. Check your key and signing method.

For the first problem we just need to add an opportunity to use Signed Headers from GET parameters.
For the second we need to add some keys to sign our canonical request.

You can see some examples bellow:

- Real S3 request
```
curl 'https://s3.amazonaws.com/{bucket}/{path}?Content-Type=text%2Fcsv&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential={accessKeyId}%2F20190520%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190520T120325Z&X-Amz-Expires=43200&X-Amz-Signature=2ec17bdeb40572d62072cc6d05741119dd9b93aa38249bb9cb0e75ded0271910&X-Amz-SignedHeaders=host%3Bx-amz-server-side-encryption&x-amz-server-side-encryption=AES256' -X PUT -H 'Accept: application/json, text/plain, */*' -H 'Referer: https://localhost:8000/' -H 'Origin: https://localhost:8000' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.131 Safari/537.36' -H 'App: sb4sf' -H 'Content-Type: text/csv' --compressed
```
Response status is 200

- Minio request
```
curl 'https://localhost:8081/{bucket}/{path}?Content-Type=text%2Fcsv&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential={accessKeyId}%2F20190520%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190520T121912Z&X-Amz-Expires=43200&X-Amz-Signature=96aaeb7b38782d52ac1a248879bc13ac5a83ac086f829880c576f5dbb42db055&X-Amz-SignedHeaders=host%3Bx-amz-server-side-encryption&x-amz-server-side-encryption=AES256' -X PUT -H 'Accept: application/json, text/plain, */*' -H 'Referer: https://localhost:8000/' -H 'Origin: https://localhost:8000' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.131 Safari/537.36' -H 'App: sb4sf' -H 'Content-Type: text/csv' --compressed
```

By default we have 400 with body:
```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>There were headers present in the request which were not signed</Message><Key>{path}</Key><BucketName>{bucket}</BucketName><Resource>/{backet}/{path}</Resource><RequestId>15A064137D73D4F8</RequestId><HostId>1aacf806-dfb5-413c-a492-42dec7fc9c47</HostId></Error>
```
Then when I've added x-amz-server-side-encryption to Signed Headers I have next error:
```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><Key>{path}</Key><BucketName>dev-burn</BucketName><Resource>/{bucket}/{path}</Resource><RequestId>15A06457C9708B00</RequestId><HostId>1aacf806-dfb5-413c-a492-42dec7fc9c47</HostId></Error>
```

## Motivation and Context

Need to fix it as well

## Regression

Need to test signature v4 algorithm

## How Has This Been Tested?

Generate pre-signed request url to upload some file.
Then send PUT request by this url as an example.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x]  I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x]  All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#5434c244-eac2-458d-aa87-67855dce1883) |
<a name="5434c244-eac2-458d-aa87-67855dce1883"></a>
### 7674-13d6fa6/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30145
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2df43846c4e0:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 50 seconds
(4/14) Running aws-sdk-ruby tests ... done in 14 seconds
(5/14) Running awscli tests ... done in 7 minutes and 5 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 31 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 11 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 59 seconds
(10/14) Running minio-java tests ... FAILED in 8 minutes and 18 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName)",
  "duration": 300397,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=XMinioReadQuorum, message=Multiple disk failures, unable to reconstruct data., bucketName=minio-java-test-308q35h, objectName=minio-java-test-3an1in6, resource=/minio-java-test-308q35h/minio-java-test-3an1in6, requestId=15A066B3B0A7C935, hostId=e9f30a79-b0ed-4b15-ac20-ef1407596e4a)\nrequest={method=GET, url=http://72.28.97.56:30145/minio-java-test-308q35h/minio-java-test-3an1in6, headers=Host: 72.28.97.56:30145\nUser-Agent: Minio (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20190520T131953Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20190520/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=503, headers=Accept-Ranges: bytes\nContent-Length: 1024\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"0fe5a5c739501e8cbc64509d7022d0cf-1\"\nLast-Modified: Mon, 20 May 2019 13:19:52 GMT\nRetry-After: 120\nServer: MinIO/DEVELOPMENT.2019-05-20T12-46-51Z\nVary: Origin\nX-Amz-Request-Id: 15A066B3B0A7C935\nX-Minio-Deployment-Id: e9f30a79-b0ed-4b15-ac20-ef1407596e4a\nX-Xss-Protection: 1; mode=block\nDate: Mon, 20 May 2019 13:19:53 GMT\n}\n >>> [io.minio.MinioClient.executeReq(MinioClient.java:1190), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executeGet(MinioClient.java:1284), io.minio.MinioClient.getObject(MinioClient.java:1860), io.minio.MinioClient.getObject(MinioClient.java:1594), FunctionalTest.copyObject_test1(FunctionalTest.java:2017), FunctionalTest.runTests(FunctionalTest.java:3006), FunctionalTest.main(FunctionalTest.java:3121)]"
}

Executed 9 out of 14 tests successfully.
```

**Comment by harshavardhana** [comment]
Thanks for the PR PTAL a look at the CI failures @mikhalenia

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#43e7d85c-dbf3-4954-89b7-1584e6c080fa) |
| mint-compression-dist-xl.sh | :x: [more...](#fb10da96-cb3c-4b03-aac8-860642d35377) |
<a name="43e7d85c-dbf3-4954-89b7-1584e6c080fa"></a>
### 7674-fa301e7/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:30283
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 12bbb741a3c8:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 20 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 46 seconds
(4/14) Running aws-sdk-ruby tests ... done in 6 seconds
(5/14) Running awscli tests ... done in 7 minutes and 14 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 2 minutes and 13 seconds
(8/14) Running minio-dotnet tests ... done in 20 minutes and 40 seconds
(9/14) Running minio-go tests ... done in 2 minutes and 20 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 34 seconds
(11/14) Running minio-js tests ... FAILED in 2 minutes and 27 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 1735,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:56:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:117:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:48:26) at DestroyableTransform.<anonymous> (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:135:12) at finishMaybe (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:371:12) at endWritable (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:378:3) at DestroyableTransform.Writable.end (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:356:5) at IncomingMessage.onend (_stream_readable.js:514:10) at endReadableNT (_stream_readable.js:978:12) at _combinedTickCallback (internal/process/next_tick.js:80:11) at process._tickCallback (internal/process/next_tick.js:104:9)"
}

Executed 10 out of 14 tests successfully.
```
<a name="fb10da96-cb3c-4b03-aac8-860642d35377"></a>
### 7674-fa301e7/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31930
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a6ab10b23178:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 20 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 4 minutes and 46 seconds
(4/14) Running aws-sdk-ruby tests ... done in 8 seconds
(5/14) Running awscli tests ... done in 5 minutes and 32 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 1 minutes and 32 seconds
{
  "name": "mc",
  "duration": "30359",
  "function": "test_put_object_0byte",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-19958 --quiet --no-color cp myminio/mc-test-bucket-714/mc-test-object-23911 mc-test-object-23911.downloaded  >>> \nmc: <ERROR> Unable to validate source myminio/mc-test-bucket-714/mc-test-object-23911"
}

Executed 6 out of 14 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

File: cmd/signature-v4-utils.go [hunk 1]
```diff
@@ -146,16 +146,22 @@ func sumHMAC(key []byte, data []byte) []byte {
...
 	extractedSignedHeaders := make(http.Header)
 	for _, header := range signedHeaders {
+		canonicalHeaderKey := http.CanonicalHeaderKey(header)
 		// `host` will not be found in the headers, can be found in r.Host.
 		// but its alway necessary that the list of signed headers containing host in it.
-		val, ok := reqHeaders[http.CanonicalHeaderKey(header)]
+		val, ok := reqHeaders[canonicalHeaderKey]
+		if !ok {
+			// try to set headers from Query String
+			val, ok = reqQueries[strings.ToLower(canonicalHeaderKey)]
```
Comment (krisis) [inline cmd/signature-v4-utils.go]: What is the reason to look for `canonicalHeaderKey` in lower-case in the `reqQueries` map? Shouldn't we look for the signed header with the same case?

**Review by krisis — CHANGES_REQUESTED** [review]

File: cmd/signature-v4-utils.go (hunk 1 above)
Comment (mikhalenia) [reply cmd/signature-v4-utils.go]: I think we needn't use CanonicalHeaderKey for this case at all. We just need to check if we can get the header value from query string. Fixed, thanks.

**Comment by mikhalenia** [comment]
> Thanks for the PR PTAL a look at the CI failures @mikhalenia

fixed

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#e245eff9-3b05-497d-9c34-565efa8db780) |
<a name="e245eff9-3b05-497d-9c34-565efa8db780"></a>
### 7674-0d47a88/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32253
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bc857a4d343e:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-gehomvlcyp30ds",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 360010,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15A0D2394CEED542, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

**Review by harshavardhana — APPROVED** [review]

**Review by krisis — APPROVED** [review]

**Pull request merged**